ambry-api/src/test/java/com/github/ambry/server/StatsSnapshotTest.java

```diff
@@ -66,5 +66,147 @@ public void deserializeStatsSnapshotTest() throws IOException {
+      }
+    });
+    StatsSnapshot copy = new StatsSnapshot(original);
+    assertTrue(copy != original);
+    assertEquals(original, copy);
+
+    assertTrue(copy.getSubMap() != original.getSubMap());
+    assertEquals(original.getSubMap(), copy.getSubMap());
+  }
+
+  /**
+   * Test {@link StatsSnapshot#updateValue()}.
+   */
+  @Test
+  public void testUpdateValue() {
+    // Anonymous class with static block initialization, not a good way to create map, it's only for map.
+    StatsSnapshot snapshot = new StatsSnapshot(0L, new HashMap<String, StatsSnapshot>() {
+      {
+        put("L1_K1", new StatsSnapshot(0L, new HashMap<String, StatsSnapshot>() {
+          {
+            put("L2_K1", new StatsSnapshot(40L, null));
+            put("L2_K2", new StatsSnapshot(60L, null));
+          }
+        }));
+        put("L1_K2", new StatsSnapshot(0L, new HashMap<String, StatsSnapshot>() {
+          {
+            put("L2_K3", new StatsSnapshot(70L, null));
+            put("L2_K4", new StatsSnapshot(90L, null));
+          }
+        }));
+        put("L1_K3", new StatsSnapshot(0L, new HashMap<String, StatsSnapshot>() {
+          {
+            put("L2_K5", new StatsSnapshot(20L, new HashMap<String, StatsSnapshot>() {
+              {
+                put("L3_K1", new StatsSnapshot(100L, null));
+                put("L3_K2", new StatsSnapshot(200L, null));
+              }
+            }));
+            put("L2_K6", new StatsSnapshot(300L, null));
+          }
+        }));
+      }
+    });
+    snapshot.updateValue();
+    assertEquals(40 + 60 + 70 + 90 + 100 + 200 + 300, snapshot.getValue());
+    assertEquals(40 + 60, snapshot.getSubMap().get("L1_K1").getValue());
+    assertEquals(70 + 90, snapshot.getSubMap().get("L1_K2").getValue());
+    assertEquals(100 + 200 + 300, snapshot.getSubMap().get("L1_K3").getValue());
+    assertEquals(100 + 200, snapshot.getSubMap().get("L1_K3").getSubMap().get("L2_K5").getValue());
+  }
+
+  /**
+   * Test {@link StatsSnapshot#removeZeroValueSnapshots()}
+   */
+  @Test
+  public void testRemoveZeroValueSnapshots() {
+    // Anonymous class with static block initialization, not a good way to create map, it's only for map.
+    StatsSnapshot snapshot = new StatsSnapshot(0L, new HashMap<String, StatsSnapshot>() {
+      {
+        put("L1_K1", new StatsSnapshot(0L, new HashMap<String, StatsSnapshot>() {
+          {
+            put("L2_K1", new StatsSnapshot(0L, null));
+            put("L2_K2", new StatsSnapshot(0L, null));
+          }
+        }));
+        put("L1_K2", new StatsSnapshot(0L, new HashMap<String, StatsSnapshot>() {
+          {
+            put("L2_K3", new StatsSnapshot(70L, null));
+            put("L2_K4", new StatsSnapshot(0L, null));
+          }
+        }));
+        put("L1_K3", new StatsSnapshot(0L, new HashMap<String, StatsSnapshot>() {
+          {
+            put("L2_K5", new StatsSnapshot(20L, new HashMap<String, StatsSnapshot>() {
```

Comment:
In practice, `L2_K5` initial value `20L` shouldn't happen, right?

Reply:
that's right, I was just being lazy to calculate the values for these snapshots so I call `updateValue` before `removeZeroValueSnapshots`. Anyway, I will put the right values in the constructor.